src/lib.rs

```diff
@@ -0,0 +1,54 @@
+#[macro_use]
+extern crate nickel;
+extern crate openssl;
+extern crate rand;
+extern crate rustc_serialize;
+extern crate byteorder;
+extern crate time;
+extern crate cookie;
+extern crate plugin;
+extern crate typemap;
+
+use std::fmt::Debug;
+use rustc_serialize::{Encodable, Decodable};
+use std::any::Any;
+use nickel::{Response, Middleware, MiddlewareResult, cookies, NickelError, Session, SessionStore};
+use nickel::status::StatusCode::Forbidden;
+
+pub type ValidateUserFunc<T> = Box<Fn(&T) -> bool + Send + Sync>;
+
+pub struct Authorizer<T, D> {
+    authorize: ValidateUserFunc<T>,
+    access_granted: Box<Middleware<D>+ Sized + Send + Sync>,
+}
+
+impl<T, D> Authorizer<T, D> {
+    pub fn new(authorize: ValidateUserFunc<T>,
+               access_granted: Box<Middleware<D> + Sized + Send + Sync>)
+               -> Authorizer<T, D> {
+        Authorizer { authorize: authorize, access_granted: access_granted }
+    }
+}
+
+impl< T, D: 'static> Middleware<D> for Authorizer<T, D>
+    where
+    D: AsRef<cookies::SecretKey> + SessionStore<Store=T>,
+    T: 'static + Any + Encodable + Decodable + Default + Debug
```

Comment:
I think the following may be along the lines we want:
```rust
trait CurrentUser : SessionStore {
   type User;

    fn current_user(&<Self as SessionStore>::Store) -> &Self::User;
}
```

And then base a trait on `Environment/Response` on that. I'm not sure where the value of adding a default implementation lies though, this behavior is very specific to the application. I think this trait is really just to extract the current user from a session so that `Permissions` can be written in terms of a user model rather than the entire `Session`.